File: list/reverse_linkedlist.cc
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


ListNode* reverse(ListNode *node) {
    ListNode* prev = nullptr;
    ListNode* nxt = nullptr;
    while (node) {
        nxt = node->next;
        node->next = prev;
        prev = node;
        node = nxt;
    }
    return prev;
}

int main() {
    int N = 10;
    ListNode* n = new ListNode(0);
    auto np = n;
    for (int i = 1; i < N; ++i) {
        np->next = new ListNode(i);
        np = np->next;
    }

    np = n;
    for (int i = 0; i < N; ++i) {
        cout << np->val << "->";
        np = np->next;
    }
    cout << endl;
    auto nn = reverse(n);

    for (int i = 0; i < N; ++i) {
        cout << nn->val << "->";
        nn = nn->next;
    }
    cout << endl;

    return 1;
}

File: list/reverse_linkedlist.go
package main

import "fmt"

type Node struct {
	val int
	nxt *Node
}

func reverse(n *Node) *Node {
	bef := (*Node)(nil)
	tmp := n
	for tmp != nil {
		next := tmp.nxt
		tmp.nxt = bef
		//next.nxt = tmp
		bef = tmp
		tmp = next
	}
	return bef
}

func main() {
	N := 10
	n := Node{0, nil}
	np := &n
	for i := 1; i < N; i++ {
		np.nxt = &Node{i, nil}
		np = np.nxt
	}
	fmt.Println()
	np = &n
	for i := 0; i < N; i++ {
		fmt.Print(np.val, "->")
		np = np.nxt
	}
	fmt.Println()
	np = &n
	rnp := reverse(np)

	for i := 0; i < N; i++ {
		fmt.Print(rnp.val, "->")
		rnp = rnp.nxt
	}
}
